internal/scan/scan.go

```diff
@@ -211,37 +198,91 @@ func (s scanner) ScanEvents(ctx context.Context, fromBlock, toBlock *big.Int) ([
 		return nil, nil
 	}
 
-	contractAbi, err := abi.JSON(strings.NewReader(contract.Erc721universalMetaData.ABI))
+	erc721UniversalAbi, err := abi.JSON(strings.NewReader(contract.Erc721universalMetaData.ABI))
 	if err != nil {
 		return nil, fmt.Errorf("error instantiating ABI: %w", err)
 	}
+
+	evoAbi, err := abi.JSON(strings.NewReader(contract.EvolutionMetaData.ABI))
+	if err != nil {
+		return nil, fmt.Errorf("error instantiating ABI: %w", err)
+	}
+
 	var parsedEvents []Event
 	for i := range eventLogs {
 		slog.Info("scanning event", "block", eventLogs[i].BlockNumber, "txHash", eventLogs[i].TxHash)
-		switch eventLogs[i].Topics[0].Hex() {
-		case eventTransferSigHash:
-			transfer, err := parseTransfer(&eventLogs[i], &contractAbi)
-			if err != nil {
-				return nil, err
-			}
-			parsedEvents = append(parsedEvents, transfer)
-			slog.Info("received event", eventTransferName, transfer)
-		case eventApprovalSigHash:
-			approval, err := parseApproval(&eventLogs[i], &contractAbi)
-			if err != nil {
-				return nil, err
-			}
-			parsedEvents = append(parsedEvents, approval)
-			slog.Info("received event", eventApprovalName, approval)
-		case eventApprovalForAllSigHash:
-			approvalForAll, err := parseApprovalForAll(&eventLogs[i], &contractAbi)
-			if err != nil {
-				return nil, err
+		if len(eventLogs[i].Topics) > 0 {
+			switch eventLogs[i].Topics[0].Hex() {
+			// Ownership events
+			case eventTransferSigHash:
+				transfer, err := parseTransfer(&eventLogs[i], &erc721UniversalAbi)
+				if err != nil {
+					if err != eventTopicsError {
+						return nil, err
+					}
+					slog.Warn("incorrect number of topics found in Transfer event",
+						"topics_found", len(eventLogs[i].Topics),
+						"topics_expected", 4)
```

Comment:
Why is this check required? If an event matches with the hash signature, it should not be possible to have fewer topics than those specified in the event declaration.

Reply:
While scanning the whole chain we found some events with less topics, or, no topics at all. As a safeguard, instead of panicking, we chose to implement the check